File: Data Structures & Algorithms/trapping-rain-water/submission-4.py
class Solution:
    # O(n) time, O(1) space
    def trap(self, height: List[int]) -> int:
        l = 0
        r = len(height) - 1

        total_water = 0
        max_left = height[l]
        max_right = height[r]
        while l < r:
            if max_left < max_right:
                l += 1
                total_water += max(0, max_left - height[l])
                max_left = max(max_left, height[l])
            else:
                r -= 1
                total_water += max(0, max_right - height[r])
                max_right = max(max_right, height[r])

        return total_water
            
        

# [0, 2, 0, 3, 1, 0, 1, 3, 2, 1]
#  l                          r
# max_left  = 0
# max_right = 1

# [0, 2, 0, 3, 1, 0, 1, 3, 2, 1]
#     l                       r
# Move l since 0 < 1
# total_water += 0 - height[i] ---> the minimum is 0 since we moved l bc l < r
# max_left  = 2
# max_right = 1

# [0, 2, 0, 3, 1, 0, 1, 3, 2, 1]
#     l                    r
# move r since 1 < 2
# total_water += 1 - height[i] ---> the minimum is 1 since we moved r bc r < l
# max_left  = 2
# max_right = 2

# [0, 2, 0, 3, 1, 0, 1, 3, 2, 1]
#        l                    r
# move l (doesn't matter)
# total_water += 2 - height[i] ---> the minimum is 2
# max_left  = 2
# max_right = 2

File: Data Structures & Algorithms/max-water-container/submission-4.cpp
class Solution {
public:
    int maxArea(vector<int>& heights) {
        int l = 0;
        int r = heights.size() - 1;

        int max = 0;
        while (l < r) {
            int area = (r - l) * std::min(heights[l], heights[r]);

            max = std::max(max, area);
            if (heights[l] < heights[r]) l += 1;
            else r -= 1;
        }

        return max;
    }
};

// [1, 7, 2, 5, 4, 7, 3, 6]
// 
// bottleneck is the shorter one
// we can move the taller one because the pointer at the shorter one is CAPPED at exactly that.
// no matter how much taller another one is it will have the same height since its shorter and
// with a smaller width

// [1, 7, 2, 5, 4, 7, 3, 6]

// 
//
// 